### PR TITLE
Results table fix, improving document circulation statistics

### LINKED ISSUE #154: Document active loan counter counts missing items as loaned
https://github.com/inveniosoftware/invenio-app-ils/blob/c4e3d7a0b55b8e9fb698369d6d4d18e4d0f6d955/invenio_app_ils/records/jsonresolver/circulation.py#L35

This line does not take into account that items might be missing and counts them as active loans

### LINKED ISSUE #173: change query for overbooked loans - dashboard
in state: ITEM_ON_LOAN, currently gets all the loans

### LINKED ISSUE #176: Showing total number does not work
The total number of records is always 10 in ResultTable component (due to elastic search config of pagination), it should take data from total field of the response

### PR DESCRIPTION
**Pull request by vlad-bm**
(Closes #173)
(Closes #176)
(Closes #154)

### CODE REVIEW

File: invenio_app_ils/records/jsonresolver/document.py [hunk 1]
```diff
@@ -27,36 +24,23 @@ def jsonresolver_loader(url_map):
...
-
-        for pid in item_pids:
-            items_count += 1
-            if not is_item_available(pid):
-                active_loans_count += 1
 
-            if circulation_is_item_available(pid):
-                loanable_items_count += 1
+        items = list(ItemSearch().search_by_document_pid(document_pid).scan())
+        items_count = len(items)
```
Comment (topless) [inline invenio_app_ils/records/jsonresolver/document.py]: `items_count` and `loanable_items_count` can be directly assigned to the relevant circulation property, to avoid the clatter :)

File: invenio_app_ils/ui/src/invenio_app_ils/pages/backoffice/Home/components/lists/OverdueLoansList/__tests__/OverdueLoansList.js [hunk 2]
```diff
@@ -105,17 +108,20 @@ describe('OverdueLoansList tests', () => {
...
-        updated: d,
-        created: d,
-        start_date: d,
-        end_date: d,
-      },
-    ];
+    const data = {
+      hits: [
+        {
+          ID: 'loan1',
```
Comment (topless) [inline invenio_app_ils/ui/src/invenio_app_ils/pages/backoffice/Home/components/lists/OverdueLoansList/__tests__/OverdueLoansList.js:114]: Is this `ID` required? I haven't seen it in any of the previous test data. The same question stands for all the occurrences further down.

**Review by topless — COMMENTED** [review]
Apart from a couple of minor things, which I might have also misunderstood, looks great! Good job!

**Review by topless — DISMISSED** [review]

File: invenio_app_ils/circulation/search.py [hunk 3]
```diff
@@ -30,6 +30,34 @@ def loan_permission_filter():
...
+        active_loans = search_by_pid(document_pid=document_pid,
+                                     filter_states=current_app.config.get(
+                                         "CIRCULATION_STATES_LOAN_ACTIVE", []
+                                     ),
+                                     ).execute().hits
+        for loan in active_loans:
+            yield loan
+
+    def get_past_loans_by_doc_pid(self, document_pid):
+        """Return any past loans for the given item."""
```
Comment (ntarocco) [inline invenio_app_ils/circulation/search.py]: Minor: given `document` instead of item?

File: invenio_app_ils/circulation/search.py [hunk 4]
```diff
@@ -30,6 +30,34 @@ def loan_permission_filter():
...
+    def get_pending_loans_by_doc_pid(self, document_pid):
+        """Return any pending loans for the given item."""
+        pending_loans = search_by_pid(document_pid=document_pid,
+                                      filter_states=['PENDING']
+                                      ).execute().hits
+        for loan in pending_loans:
+            yield loan
+
+    def get_active_loans_by_doc_pid(self, document_pid):
+        """Return any active loans for the given item."""
```
Comment (ntarocco) [inline invenio_app_ils/circulation/search.py]: Same here

File: invenio_app_ils/circulation/search.py [hunk 5]
```diff
@@ -30,6 +30,34 @@ def loan_permission_filter():
 class IlsLoansSearch(LoansSearch):
     """LoanSearch to filter loans."""
 
+    def get_pending_loans_by_doc_pid(self, document_pid):
+        """Return any pending loans for the given item."""
```
Comment (ntarocco) [inline invenio_app_ils/circulation/search.py]: Same here

File: invenio_app_ils/records/jsonresolver/document.py [hunk 6]
```diff
@@ -27,36 +24,23 @@ def jsonresolver_loader(url_map):
...
-            filter_states=current_app.config.get(
-                "CIRCULATION_STATES_LOAN_COMPLETED", []
-            ),
-        )
+        loan_search = IlsLoansSearch()
 
-        pending_loans = search_by_pid(document_pid=document_pid,
-                                      filter_states=['PENDING']
-                                      ).execute().hits.total
+        past_loans_count = len(list(loan_search.get_past_loans_by_doc_pid(document_pid)))
```
Comment (ntarocco) [inline invenio_app_ils/records/jsonresolver/document.py]: You are going to hate me, and I am sorry, but it came to my mind only now. We can even improve this more by avoiding `list` and then `len`. We can ask directly to ES the count.
See [here](https://github.com/inveniosoftware/invenio-circulation/blob/792f5ae7352b7979717c8087e142f3eabe70d0e8/invenio_circulation/search/api.py#L52) and [here](https://github.com/inveniosoftware/invenio-circulation/blob/792f5ae7352b7979717c8087e142f3eabe70d0e8/invenio_circulation/api.py#L69) an example.

Basically we "prepare" the search, but it is when we use it we decide to use `execute` or `scan`.
It should not be a big effort to change it... Sorry...

File: invenio_app_ils/ui/src/invenio_app_ils/common/api/documents/document.js [hunk 7]
```diff
@@ -45,7 +45,8 @@ const queryBuilder = () => {
 
 const list = query => {
   return http.get(`${documentURL}?q=${query}`).then(response => {
-    response.data = response.data.hits.hits.map(hit =>
+    response.data.total = response.data.hits.total;
```
Comment (ntarocco) [inline invenio_app_ils/ui/src/invenio_app_ils/common/api/documents/document.js]: thanks!

**Review by ntarocco — DISMISSED** [review]

**Comment by ntarocco** [comment]
@kprzerwa if you agree we can merge!

**Review by ntarocco — DISMISSED** [review]

File: invenio_app_ils/records/jsonresolver/document.py [hunk 8]
```diff
@@ -27,36 +25,20 @@ def jsonresolver_loader(url_map):
...
-        active_loans_count = 0
-        loanable_items_count = 0
-        items_count = 0
-        item_pids = circulation_items_retriever(document_pid)
 
-        for pid in item_pids:
-            items_count += 1
-            if not is_item_available(pid):
-                active_loans_count += 1
+        items = list(ItemSearch().search_by_document_pid(document_pid).scan())
```
Comment (kpsherva) [inline invenio_app_ils/records/jsonresolver/document.py]: we could also replace this with `.hits.total`

File: invenio_app_ils/records/jsonresolver/document.py [hunk 9]
```diff
@@ -27,36 +25,20 @@ def jsonresolver_loader(url_map):
...
-
-        circulation["number_of_past_loans"] = search.execute().hits.total
-        circulation["number_of_items"] = items_count
+        circulation["number_of_past_loans"] = past_loans_count
+        circulation["number_of_items"] = len(items)
         circulation["active_loans"] = active_loans_count
-        circulation["loanable_items"] = loanable_items_count
-        circulation["pending_loans"] = pending_loans
-        circulation["overbooked"] = pending_loans > loanable_items_count
+        circulation["loanable_items"] = sum(item["status"] == "LOANABLE" for item in items)
```
Comment (kpsherva) [inline invenio_app_ils/records/jsonresolver/document.py]: the logic of this statistic is a bit wrong, because we want to count the items which are not on loan currently also, not only `LOANABLE`, that's why there was `is_item_available` function

**Review by zzacharo — CHANGES_REQUESTED** [review]

File: invenio_app_ils/circulation/search.py [hunk 10]
```diff
@@ -33,23 +33,22 @@ class IlsLoansSearch(LoansSearch):
...
+                             filter_states=["PENDING"]
                              )
 
     def get_active_loans_by_doc_pid(self, document_pid):
         """Return any active loans for the given document."""
         return search_by_pid(document_pid=document_pid,
-                             filter_states=current_app.config.get(
-                                 "CIRCULATION_STATES_LOAN_ACTIVE", []
-                             ),
+                             filter_states=["ITEM_AT_DESK",
```
[the remark this answers is not in the document]
Comment (vlad-bm) [reply invenio_app_ils/circulation/search.py]: My mistake, I did not know it could get the config states from invenio circulation, I changed them back

**Review by zzacharo — DISMISSED** [review]
